sbat.c

```diff
@@ -335,4 +335,75 @@ parse_sbat_var(list_t *entries)
 	FreePool(data);
 	return EFI_INVALID_PARAMETER;
 }
+
+EFI_STATUS
+set_sbat_uefi_variable(void)
+{
+	EFI_STATUS efi_status = EFI_SUCCESS;
+	UINT32 attributes = 0;
+
+	CHAR8 *sbat = NULL;
+	UINTN sbatsize = 0;
+
+	/* get size */
+	efi_status = gRT->GetVariable(L"SBAT", &SHIM_LOCK_GUID, &attributes,
+			&sbatsize, sbat);
+	sbat = AllocateZeroPool(sbatsize);
+	/* read and get attributes */
+	efi_status = gRT->GetVariable(L"SBAT", &SHIM_LOCK_GUID, &attributes,
+			&sbatsize, sbat);
+
+	/*
+	 * If attributes and the initial part of the version match, then
+	 * there is no need to re-write it. If it is not set or the read
+	 * failed, then attributes won't match.
+	 *
+	 * Finally, if the SBAT_VAR built into this shim is newer, update
+	 * it. Either the sbat version or date can be newer, and since
+	 * the version is more significant than the date, strcmp() applies.
+	 */
+	if (((attributes != (EFI_VARIABLE_NON_VOLATILE |
+		EFI_VARIABLE_BOOTSERVICE_ACCESS))
+		|| (attributes != (EFI_VARIABLE_NON_VOLATILE |
+				EFI_VARIABLE_BOOTSERVICE_ACCESS |
+				EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS)))
+		|| (strncmp(sbat, SBAT_VAR_SIG, strlen(SBAT_VAR_SIG)) != 0)
+		|| (strncmp(SBAT_VAR, sbat, strlen(SBAT_VAR)) > 0)
+		) {
+
+		/* delete previous variable */
+		efi_status = gRT->SetVariable(L"SBAT", &SHIM_LOCK_GUID,
+				attributes, 0, "");
+
+		/* verify that it's gone */
+		efi_status = gRT->GetVariable(L"SBAT", &SHIM_LOCK_GUID,
+				0, &sbatsize, sbat);
+		if ((efi_status != EFI_SUCCESS) || (sbatsize != 0))
+			dprint(L"SBAT variable clearing failed\n");
+
+		/* set variable */
+		efi_status = gRT->SetVariable(L"SBAT", &SHIM_LOCK_GUID,
+				EFI_VARIABLE_NON_VOLATILE |
+				EFI_VARIABLE_BOOTSERVICE_ACCESS,
+				sizeof (SBAT_VAR), SBAT_VAR);
+
+		/* verify that the expected data is there */
+		FreePool(sbat);
+		efi_status = gRT->GetVariable(L"SBAT", &SHIM_LOCK_GUID,
+				0, &sbatsize, NULL);
+		sbat = AllocateZeroPool(sbatsize);
```

Comment:
Do we need to free and allocate the buffer if we know that the data will have the same size? In any case you can check if the `sbatsize` doesn't match the expected one and report an error.

Reply:
I think that you could just `ZeroMem(sbat)` instead of the free + alloc.

Reply:
You are actually correct because the `sbatsize` is the size of the original read variable while the second allocation should match `strlen(SBAT_VAR)`.

Probably I would add a check since there's no point to compare sbat and SBAT_VAR if the sizes don't match.